File: lclang/src/main/java/lclang/LCClass.java
package lclang;

import lclang.exceptions.LCLangRuntimeException;
import lclang.libs.lang.classes.StringClass;
import lclang.methods.Method;
import lclang.types.NamedType;
import org.jetbrains.annotations.Nullable;

import java.util.*;

public class LCClass extends LCRootExecutor {
    public static int classesCount = 0;

    public final String name;
    public final int classId = classesCount++;
    public LCClass extendsClass;
    public Method constructor = null;

    public LCClass(String name) {
        this(name, name);
    }

    public LCClass(String name, String path) {
        super(path);
        this.name = name;
    }

    @Nullable
    @Override
    public Value getLink(Caller caller, String name) throws LCLangRuntimeException {
        Value value = super.getLink(caller, name);
        if(value==null&&extendsClass!=null)
            return extendsClass.executor.getLink(caller, name);

        return value;
    }

    public Value asValue(){
        return new Value(new NamedType(this), this, Value.State.NOTHING);
    }

    public boolean canCast(LCClass another) {
        return Objects.equals(another.name, name) ||
            extendsClass!=null && extendsClass.canCast(another);
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof LCClass && classId == other.hashCode();
    }

    public String toString(Caller caller) throws LCLangRuntimeException {
        Value toStringMethod = globals.get("toString");
        if(toStringMethod!=null) {
            LCClass clazz = toStringMethod.get.invoke(caller);
            if(clazz instanceof Method) {
                Method method = (Method) clazz;
                return ((StringClass) method.call(caller, Collections.emptyList()).get.invoke(caller)).string;
            }
        }

        List<String> parameters = new ArrayList<>();
        Set<Map.Entry<String, Value>> entries = executor.variables.entrySet();
        for (Map.Entry<String, Value> entry : entries) {
            parameters.add(entry.getKey()+": "+entry.getValue().type+" = "+
                    entry.getValue().get.invoke(caller));
        }

        return name+"@class"+classId+": { " + String.join(", ", parameters) + " }";
    }

    @Override
    @Deprecated
    public String toString() {
        try {
            return toString(new Caller(this, 0, null));
        }catch (Exception e){
            throw new RuntimeException(e);
        }
    }

    @Override
    public int hashCode() {
        return classId;
    }
}

File: lclang/src/main/kotlin/lclang/LCClass.kt
package lclang

import lclang.methods.Method
import lclang.types.NamedType

open class LCClass(
    val name: String,
    path: String = name
): LCRootExecutor(path) {
    constructor(name: String, root: LCRootExecutor): this(name, root.path)

    private val classId = Global.classIndex++
    open var extends: LCClass? = null
    open var constructor: Method? = null

    open fun asValue() = Value(NamedType(this), this)

//    companion object {
//        fun from(componentName: String,
//                 root: LCRootExecutor,
//                 fromClass: lclangParser.ClassExprContext): LCClass {
//            val staticClass = LCClass(componentName+fromClass.name.text, root)
//            val clazzType = NamedType(staticClass)
//            staticClass.constructor = VisitorMethod(root.executor, clazzType,
//                fromClass.args()?.arg()?:listOf()) {
//                val clazz = LCClass(staticClass.name, root)
//                if(fromClass.classExtends!=null) {
//                    clazz.extends = root.classes[fromClass.classExtends.text] ?: throw ClassNotFoundException(
//                        staticClass.name,
//                        fromClass.classExtends.asCaller(root)
//                    )
//
//                    clazz.accept(clazz.extends!!)
//                }
//                clazz.executor.variables.putAll(it.variables)
//                clazz.accept(root)
//
//                for(method in fromClass.method())
//                    clazz.globals[method.ID().text] = ClassMethod(clazz, method).asValue()
//
//                for (stmt in fromClass.stmt()) clazz.compiler.visitStmt(stmt)
//                    .visit(Caller(clazz, 0, 0), clazz.executor)
//
//                return@VisitorMethod Value(clazzType, clazz)
//            }
//
//            return staticClass
//        }
//    }

    fun canCast(another: LCClass): Boolean =
        another.name == name || extends!=null && extends!!.canCast(another)

    override fun equals(other: Any?) = other is LCClass && classId == other.classId
    override fun hashCode(): Int = classId
    override fun toString(): String {
        val caller = Caller(this, 0, 0)

        val toStringMethod = globals["toString"]
        if(toStringMethod!=null) {
            val method = toStringMethod.get(caller)
            if(method is Method)
                return method.call(caller, listOf()).toString()
        }

        return "$name@class$classId: {" + executor.variables.entries.joinToString {
            it.key + ": " + it.value.type + " = " + it.value.get(caller)
        } + "}"
    }
}